docs/learn.md

```diff
@@ -721,20 +723,1250 @@ template](https://github.com/melange-re/melange-opam-template).
+arguments need to belong to the same type.
+
+```ocaml
+external join : string array -> string = "join"
+  [@@bs.module "path"] [@@bs.splice]
+let v = join [| "a"; "b" |]
+```
+
+Generates:
+
+```js
+var Path = require("path");
+var v = Path.join("a", "b");
+```
+
+#### Bing to a polymorphic function
+
+Some JavaScript libraries will define functions that where the arguments can
+vary on both type and shape.There are two approaches to bind to those, depending
+on how dynamic they are.
+
+##### Approach 1: Multiple external functions
+
+If it is possible to enumerate the many forms an overloaded JavaScript function
+can take, a flexible approach is to bind to each form individually:
+
+```ocaml
+external drawCat : unit -> unit = "draw" [@@bs.module "MyGame"]
+external drawDog : giveName:string -> unit = "draw" [@@bs.module "MyGame"]
+external draw : string -> useRandomAnimal:bool -> unit = "draw"
+  [@@bs.module "MyGame"]
+```
+
+Note how all three externals bind to the same JavaScript function, `draw`.
+
+##### Approach 2: Polymorphic variant + `bs.unwrap`
+
+In some cases, the function has a constant number of arguments but the type of
+the argument can vary. For cases like this, we can model the argument as a
+variant and use the `bs.unwrap` attribute in the external.
+
+Let’s say we want to bind to the following JavaScript function:
+
+```js
+function padLeft(value, padding) {
+  if (typeof padding === "number") {
+    return Array(padding + 1).join(" ") + value;
+  }
+  if (typeof padding === "string") {
+    return padding + value;
+  }
+  throw new Error(`Expected string or number, got '${padding}'.`);
+}
+```
+
+Here, `padding` can be modeled as a polymorphic variant.
+
+```ocaml
+external padLeft:
+  string
+  -> ([ `Str of string
+      | `Int of int
+      ] [@bs.unwrap])
+  -> string
+  = "padLeft" [@@bs.val]
+
+let _ = padLeft "Hello World" (`Int 4)
+let _ = padLeft "Hello World" (`Str "Message from Melange: ")
+```
+
+Which produces the following JavaScript:
+
+```js
+padLeft("Hello World", 4);
+padLeft("Hello World", "Message from Melange: ");
+```
+
+As we saw in the [Non-shared data types](#non-shared-data-types) section, we
+should rather avoid passing variants directly to the JavaScript side. By using
+`bs.unwrap` we get the best of both worlds: from Melange we can use variants,
+while JavaScript gets the raw values inside them.
+
+#### Using polymorphic variants to bind to enums and string types
+
+Some JavaScript APIs take a limited subset of values as input. For example,
+Node’s `fs.readFileSync` second argument. It can only take a few given string
+values: `"ascii"`, `"utf8"`, etc.
+
+One could still bind it just as a string, but this would not prevent from using
+unsupported values. Let’s see how we can use polymorphic variants and the
+`bs.string` attribute to avoid runtime errors:
+
+```ocaml
+external read_file_sync :
+  name:string -> ([ `utf8 | `ascii ][@bs.string]) -> string = "readFileSync"
```

Comment:
I think this is not needed anymore since polyvariants compile to strings.

Reply:
I tried removing `bs.string` but it seems it makes using `bs.as` impossible, tracked in https://github.com/melange-re/melange/issues/578. So for now decided to leave it.

Added an example with `bs.string` and `bs.as` in https://github.com/melange-re/documentation-site/pull/12/commits/15499fb1f3e577830b4fc6e3a4e8c29d28d51b15.